Receiving events via Kafka and Publishing to WebSocket (#3477)
wso2/docs-apim: en/docs/use-cases/streaming-tutorials/receiving-events-via-kafka-and-publishing-to-websocket.md

# Recieving XML events via Kafka and Publishing to WebSocket

## Introduction

> The Streaming Integrator can consume events from a Kafka topic and publish those events to a WebSocket  in a streaming manner.

Reviewer: ```suggestion
The Streaming Integrator can consume events from a Kafka topic and publish those events to a WebSocket in a streaming manner.
```

Shall we expand what we mean by "streaming manner"?

1. Download the Kafka broker from [the Apache site](https://www.apache.org/dyn/closer.cgi?path=/kafka/2.3.0/kafka_2.12-2.3.0.tgz) and extract it. 
From here onwards, this directory is referred to as `<KAFKA_HOME>`.

2. Create a directory named `Source` in a preferred location in your machine and copy the following JARs to the `Source` directory from the `<KAFKA_HOME>/libs` directory.

    - `kafka_2.12-2.3.0.jar`
   
    - `kafka-clients-2.3.0.jar`
    
    - `metrics-core-2.2.0.jar`
    
    - `scala-library-2.12.8.jar`
    
    - `zkclient-0.11.jar`
   
    - `zookeeper-3.4.14.jar`
  
3. Create another directory named `Destination` in a preferred location in your machine.

4. To convert the Kafka JARS you copied to the `Source` directory, issue the following command:

            * For Windows:
                ```
                {WSO2SIHome}/bin/jartobundle.bat <{Source} Directory Path> <{Destination} Directory Path>
                ```
            * For Linux:
                ```
                sh {WSO2SIHome}/bin/jartobundle.sh <{Source} Directory Path> <{Destination} Directory Path>
                ```
                

> 5. Add the OSGI converted kafka libs from the `Destination` directory to the `<SI_HOME>/lib` directory.

Reviewer: ```suggestion
5. Add the OSGI converted Kafka libs from the `Destination` directory to the `<SI_HOME>/lib` directory.
```

6. Add the original Kafka libs from `Source` to `<SIHome>/samples/sample-clients/lib`.

## Consuming data from Kafka

### Step 1: Start Kafka

> 1. Navigate to the `<KAFKA_HOME>` directory and start a zookeeper node by issuing the following command.

Reviewer: ```suggestion
1. Navigate to the `<KAFKA_HOME>` directory and start a Zookeeper node by issuing the following command.
```

####
1. Let's create a basic Siddhi application to consume messages from a Kafka topic and publish to WebSocket in XML format.

```
@App:name("RecieveKafkaPublishWebSocket")
@App:description("Description of the plan")

@source(type='kafka',
        topic.list='kafka_sample_topic',
        partition.no.list='0',
        threading.option='single.thread',
        group.id='group',
        bootstrap.servers='localhost:9092',
@map(type='xml'))
define stream SweetProductionStream(name string, amount double);

@sink(type='websocket', 
      url='ws://localhost:8025/abc',
@map(type='xml'))
define stream LowProductionAlertStream (name string, amount double);

@info(name='query1')
from SweetProductionStream
select *
insert into LowProductionAlertStream;
```

2. Save the sample. If there is no syntax error in the Siddhi Application, the following message is shown on the console:

```
Siddhi App RecieveKafkaPublishWebSocket.siddhi successfully deployed. 
```

#### Executing the Sample:

1. Open a terminal and navigate to `<SIHome>/samples/sample-clients/kafka-producer` and run "ant" command as follows:
```
ant -Dtype=xml -DtopicName=kafka_sample_topic
```

> You can limit the number of events as follows as well,

Reviewer: Does the number formatting break here?

Author: No.

```
ant -Dtype=xml -DnoOfEventsToSend=5 -DtopicName=kafka_sample_topic
```

> 2. Open a terminal and navigate to the `<SIHome>/samples/sample-clients/websocket-receiver` directory and run the ant command.

Reviewer: ```suggestion
2. Open a terminal and navigate to the `<SIHome>/samples/sample-clients/websocket-receiver` directory and run the `ant` command.
```

    If you use the default host 'localhost' and port '8025' in your program use `ant` command without any arguments.
    However, if you use different host or port, run the ant command with appropriate arguments. 
    e.g., `ant -Dport=9025`
 

3. Start the Siddhi application by clicking on 'Run'.

4. If the Siddhi application starts successfully, the following messages would be shown on the console.
        ```
        * RecieveKafkaPublishWebSocket.siddhi -  Started Successfully!
        ```

#### Notes:

If you edit this application while it's running, stop the application -> Save -> Start.

## Viewing the Results:

> Check the output in the terminal of <SIHome>/samples/sample-clients/websocket-receiver. You will see output similar to the following: 

Reviewer: ```suggestion
Check the output in the terminal of `<SI_HOME>/samples/sample-clients/websocket-receiver`. You will see output similar to the following: 
```

Here you can view and search for all the deployed services from this interface of the Service Catalogue. To search for services, click on the search icon on the top right corner of the listing table shown in the diagram below.
 
The `SweetProdApp` that was deployed will be shown in this interface as follows.

![Service Catalogue Entry in API Manager](https://github.com/wso2/docs-apim/blob/master/en/docs/assets/img/streaming/working-with-async-api/service-catalogue-entry.png)

### Step 8: Manage Services in the Service Catalogue

View information of a specific Service in the Service Catalogue by clicking on the name of the **Service** from the listing page.

Click on **Hamburger Icon** -> Select **Service Catalog** -> Click on the **Service name**

You will be directed to the **Service Overview** page as follows where you can view service information such as,

- Service name 
- Version
- Description
- Usages in APIs
- Other important metadata.

### Step 9: Creation of an API

Create an API from a deployed service by clicking on the **Create API** button in the services listing page.

> Provide the details in there for API Name, Context and Version where all fields are mandatory, and click on the **Create API** button. 

Reviewer 2: Add an image of the entered details
